pkg/addon/addon.go

```diff
@@ -1072,6 +1092,210 @@ func (h *Installer) installDependency(addon *InstallPackage) error {
+	for _, dep := range dependenciesNotInstalled {
+		availableAddon, ok := availableAddons[dep.Name]
+		if !ok {
+			err = multierr.Append(err, fmt.Errorf(
+				"addon %s has unresolvable dependency %s, required version '%s', no available addon with name",
+				addon.Name, dep.Name, dep.Version))
+		} else {
+			_, err2 := calculateDependencyVersionToInstall(dep, nil, availableAddons)
+			if err2 != nil {
+				err = multierr.Append(err, fmt.Errorf(
+					"addon %s has unresolvable dependency %s, required version '%s', available versions %v",
+					addon.Name, dep.Name, dep.Version, availableAddon.AvailableVersions))
+			}
+		}
+	}
+
+	return err
+}
+
+// calculateDependencyVersionToInstall compares an addon's dependency to a list
+// of installed and available addons and returns a version to install.
+// If the addon is already installed, return the installed version. If the addon
+// is not installed, return the latest available version that satisfies the
+// dependency version.
+func calculateDependencyVersionToInstall(dependency Dependency, installedAddons addonInfoMap, availableAddons addonInfoMap) (string, error) {
+	if dependency.Name == "" {
+		return "", fmt.Errorf("dependency name cannot be empty")
+	}
+
+	// if dependency is already installed, return the installed version
+	if installedAddons != nil {
+		installedAddon, ok := installedAddons[dependency.Name]
+		if ok {
+			// versions length must be 1
+			if len(installedAddon.AvailableVersions) != 1 {
+				return "", errors.New("installedAddon.Versions length must be 1")
+			}
+			return installedAddon.AvailableVersions[0], nil
+		}
+	}
+
+	availableAddon, ok := availableAddons[dependency.Name]
+	if !ok {
+		return "", fmt.Errorf("no available addon with name %s", dependency.Name)
+	}
+
+	sortedVersions := sortVersionsDescending(availableAddon.AvailableVersions)
+
+	// if no version is specified, return the latest version
+	if dependency.Version == "" {
+		return sortedVersions[0], nil
+	}
+
+	// check if the dependency version is satisfied
+	var match bool
+	for _, version := range sortedVersions {
+		match, _ = checkSemVer(version, dependency.Version)
+		if match {
+			return version, nil
+		}
+	}
+	return "", fmt.Errorf("no available addon with name %s and version '%s'", dependency.Name, dependency.Version)
+}
+
+func sortVersionsDescending(versions []string) []string {
+	var sortedVersions []*semver.Version
+	var sortedVersionStrings []string
+	for _, v := range versions {
+		var err error
+		parsedVersion, err := semver.NewVersion(v)
+		if err == nil {
+			sortedVersions = append(sortedVersions, parsedVersion)
+		}
+	}
+	// sort versions in descending order
+	sort.Sort(sort.Reverse(semver.Collection(sortedVersions)))
+	for _, v := range sortedVersions {
+		sortedVersionStrings = append(sortedVersionStrings, v.String())
+	}
+	return sortedVersionStrings
+}
+
+// AddonInfo contains summary information about an addon
+type AddonInfo struct {
```

Comment:
I think the github review is valid. Should we call it addon.Info with the package prefix explicitly showing this is for an addon?

Reply:
I'll resolve this conversation for now since the lint check is resolved.

Reply:
I eventually decided that this type should be exported, since I added a method to Registry that returns this, so I renamed this to ItemInfo.